source/extensions/filters/http/cache/cache_insert_queue.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <deque>
+#include <functional>
+
+#include "source/extensions/filters/http/cache/http_cache.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+using OverHighWatermarkCallback = std::function<void()>;
+using UnderLowWatermarkCallback = std::function<void()>;
+using AbortInsertCallback = std::function<void()>;
+class CacheInsertChunk;
+
+class CacheInsertQueue {
```

Comment:
I think having a high level comment about this tracking watermarks, making sure one chunk is in flight at at time etc would be awesome